### PR TITLE
optimizer: add a batch after reduce -every

### PR DESCRIPTION
**Pull request by davidvgalbraith**
Total hack! Best way to solve this problem though if we insist on the marks thing.

@demmer @dmajda @bkutil 

fixes https://github.com/juttle/juttle/issues/484

### CODE REVIEW

**Comment by bkutil** [comment]
Q: wouldn't it be possible for the adapter, using the return value from its `read()` method, to tell the `read` that the output is batched and let `read` or `source` insert marks after each point? It might be cleaner than modifying the flow graph.

**Comment by davidvgalbraith** [comment]
Yeah! Would prefer to do magic and have this be totally opaque to the adapters.

File: lib/graph.js [hunk 1]
```diff
@@ -11,8 +11,8 @@ class Graph {
         return this.built_graph.inputs;
     }
 
-    add_node(type, name) {
-        var index = this.get_nodes().length;
+    add_node(type, name, index) {
```
Comment (dmajda) [inline lib/graph.js]: I’d slightly prefer to have a separate `insert_node(type, name, index)` method and to reimplement `add_node(type, name)` in terms of it. The idea is that appending and inserting at a specific position are two different things so they should be serviced by two different methods.

**Comment by dmajda** [comment]
+1

Once #53 is resolved I think we should consider reverting this hack and using the same/similar method we’ll use to add `-every` to various other procs.

**Pull request merged**